// @ant-design/pro-form/lib/components/Radio/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/radio/style");

var _radio = _interopRequireDefault(require("antd/es/radio"));

var _react = _interopRequireDefault(require("react"));

var _BaseForm = require("../../BaseForm");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var RadioGroup = _radio.default.Group;

var Group = _react.default.forwardRef(function (_ref, ref) {
  var children = _ref.children,
      fieldProps = _ref.fieldProps,
      options = _ref.options;

  var renderChildren = function renderChildren() {
    if (options) {
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, options.map(function (option) {
        if (typeof option === 'string') {
          return /*#__PURE__*/_react.default.createElement(_radio.default, {
            key: option,
            value: option
          }, option);
        }

        return /*#__PURE__*/_react.default.createElement(_radio.default, {
          disabled: option.disable,
          key: option.value,
          value: option.value
        }, option.label);
      }), children);
    }

    return children;
  };

  return /*#__PURE__*/_react.default.createElement(RadioGroup, _extends({
    ref: ref
  }, fieldProps), renderChildren());
});
/**
 * Radio
 * @param
 */


var ProFormRadio = _react.default.forwardRef(function (_ref2, ref) {
  var fieldProps = _ref2.fieldProps;
  return /*#__PURE__*/_react.default.createElement(_radio.default, _extends({}, fieldProps, {
    ref: ref
  }));
}); // @ts-expect-error


var WrappedProFormRadio = (0, _BaseForm.createField)(ProFormRadio, {
  valuePropName: 'checked',
  ignoreFelidWidth: true
});
WrappedProFormRadio.Group = (0, _BaseForm.createField)(Group, {
  customLightMode: true,
  ignoreFelidWidth: true
});
WrappedProFormRadio.Button = _radio.default.Button;
var _default = WrappedProFormRadio;
exports.default = _default;
