// @ant-design/pro-table/es/Form/index.js
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";
import "antd/es/form/style";
import _Form from "antd/es/form";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { useContext, useEffect, useRef, useCallback } from 'react';
import { useIntl } from '@ant-design/pro-provider';
import ProForm, { QueryFilter, LightFilter, ProFormField } from '@ant-design/pro-form';
import classNames from 'classnames';
import warningOnce from 'rc-util/lib/warning';
import omit from 'omit.js';
import { useDeepCompareEffect, conversionSubmitValue, transformKeySubmitValue } from '@ant-design/pro-utils';
import { genColumnKey } from '../utils';
import Container from '../container';
import './index.less';
/**
 *  获取当前选择的 Form Layout 配置
 * @param isForm
 * @param searchConfig
 * @returns LightFilter | QueryFilter | ProForm
 */

var getFormCompetent = function getFormCompetent(isForm, searchConfig) {
  if (!isForm && searchConfig !== false) {
    if ((searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.filterType) === 'light') {
      return {
        Competent: LightFilter,
        competentName: 'light-filter'
      };
    }

    return {
      Competent: QueryFilter,
      competentName: 'query-filter'
    };
  }

  return {
    Competent: ProForm,
    competentName: 'form'
  };
};
/**
 * 获取需要传给相应表单的props
 * @param searchConfig
 * @param name
 */


var getFromProps = function getFromProps(isForm, searchConfig, name) {
  if (!isForm && name === 'light-filter') {
    // 传给 lightFilter 的问题
    return omit(_objectSpread({}, searchConfig), ['labelWidth', 'defaultCollapsed', 'filterType']);
  }

  if (!isForm) {
    // 传给 QueryFilter 的配置
    return omit(_objectSpread({
      labelWidth: searchConfig ? searchConfig === null || searchConfig === void 0 ? void 0 : searchConfig.labelWidth : undefined,
      defaultCollapsed: true
    }, searchConfig), ['filterType']);
  }

  return {};
};

export var formInputRender = function formInputRender(props, ref) {
  var item = props.item,
      intl = props.intl,
      form = props.form,
      type = props.type,
      formItemProps = props.formItemProps,
      rest = _objectWithoutProperties(props, ["item", "intl", "form", "type", "formItemProps"]);

  var _item$valueType = item.valueType,
      itemValueType = _item$valueType === void 0 ? 'text' : _item$valueType; // if function， run it

  var valueType = (typeof itemValueType === 'function' ? itemValueType({}, type) : itemValueType) || 'text';
  /**
   * 自定义 render
   */

  if (item.renderFormItem) {
    /**
     *删除 renderFormItem 防止重复的 dom 渲染
     */
    var renderFormItem = item.renderFormItem,
        restItem = _objectWithoutProperties(item, ["renderFormItem"]);

    var defaultRender = function defaultRender(newItem) {
      return formInputRender(_objectSpread({}, _objectSpread(_objectSpread({}, props), {}, {
        item: newItem
      })));
    }; // 自动注入 onChange 和 value，用户自己很有可能忘记


    var dom = renderFormItem(restItem, _objectSpread(_objectSpread({}, rest), {}, {
      type: type,
      defaultRender: defaultRender
    }), form); // 有可能不是一个组件

    if (!React.isValidElement(dom)) {
      return dom;
    }

    var defaultProps = dom.props;

    if (defaultProps.isDefaultDom) {
      return dom;
    } // 已用户的为主，不然过于 magic


    return /*#__PURE__*/React.createElement(ProFormField, _extends({
      key: "".concat(item.dataIndex || '', "-").concat(item.key || '', "-").concat(item.index)
    }, rest, {
      ref: ref,
      initialValue: item.initialValue,
      name: item.key || item.dataIndex
    }), React.cloneElement(dom, _objectSpread(_objectSpread({}, rest), defaultProps)));
  }

  var _ref = item.fieldProps || {},
      onChange = _ref.onChange,
      restFieldProps = _objectWithoutProperties(_ref, ["onChange"]);

  var finalValueType = !valueType || ['textarea', 'jsonCode', 'code'].includes(valueType) && type === 'table' ? 'text' : valueType;
  return /*#__PURE__*/React.createElement(ProFormField, _extends({
    ref: ref,
    tooltip: item.tooltip || item.tip,
    isDefaultDom: true,
    valueEnum: item.valueEnum,
    name: item.key || item.dataIndex,
    onChange: onChange,
    fieldProps: restFieldProps || item.formItemProps // valueType = textarea，但是在 查询表单这里，应该是个 input 框
    ,
    valueType: finalValueType,
    initialValue: item.initialValue
  }, rest, {
    rules: type === 'form' ? rest.rules : undefined,
    key: "".concat(item.dataIndex || '', "-").concat(item.key || '', "-").concat(item.index)
  }));
};
export var proFormItemRender = function proFormItemRender(_ref2) {
  var item = _ref2.item,
      intl = _ref2.intl,
      formInstance = _ref2.formInstance,
      type = _ref2.type;

  var valueType = item.valueType,
      dataIndex = item.dataIndex,
      valueEnum = item.valueEnum,
      renderFormItem = item.renderFormItem,
      render = item.render,
      hideInForm = item.hideInForm,
      hideInSearch = item.hideInSearch,
      search = item.search,
      hideInTable = item.hideInTable,
      renderText = item.renderText,
      order = item.order,
      width = item.width,
      sorter = item.sorter,
      formItemProps = item.formItemProps,
      initialValue = item.initialValue,
      ellipsis = item.ellipsis,
      index = item.index,
      filters = item.filters,
      rest = _objectWithoutProperties(item, ["valueType", "dataIndex", "valueEnum", "renderFormItem", "render", "hideInForm", "hideInSearch", "search", "hideInTable", "renderText", "order", "width", "sorter", "formItemProps", "initialValue", "ellipsis", "index", "filters"]); // 支持 function 的 title


  var getTitle = function getTitle() {
    if (rest.title && typeof rest.title === 'function') {
      return rest.title(item, 'form', '');
    }

    return rest.title;
  };

  var dom = formInputRender(_objectSpread({
    item: item,
    type: type,
    intl: intl,
    form: formInstance,
    label: getTitle()
  }, formItemProps));

  if (!dom) {
    return null;
  }

  return dom;
};

var FormSearch = function FormSearch(_ref3) {
  var _classNames;

  var onSubmit = _ref3.onSubmit,
      formRef = _ref3.formRef,
      _ref3$dateFormatter = _ref3.dateFormatter,
      dateFormatter = _ref3$dateFormatter === void 0 ? 'string' : _ref3$dateFormatter,
      type = _ref3.type,
      _onReset = _ref3.onReset,
      submitButtonLoading = _ref3.submitButtonLoading,
      searchConfig = _ref3.search,
      _ref3$form = _ref3.form,
      formConfig = _ref3$form === void 0 ? {} : _ref3$form;

  /**
   * 为了支持 dom 的消失，支持了这个 api
   */
  var intl = useIntl();

  var _Form$useForm = _Form.useForm(),
      _Form$useForm2 = _slicedToArray(_Form$useForm, 1),
      form = _Form$useForm2[0];

  var formInstanceRef = useRef(form);
  var counter = Container.useContainer();
  /**
   * 保存 valueTypeRef，用于分辨是用什么方式格式化数据
   */

  var valueTypeRef = useRef();
  /**
   * 保存 transformKeyRef，用于对表单key transform
   */

  var transformKeyRef = useRef({}); // 这么做是为了在用户修改了输入的时候触发一下子节点的render

  var _React$useState = React.useState(),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      updateState = _React$useState2[1];

  var forceUpdate = useCallback(function () {
    return updateState(undefined);
  }, []);
  var isForm = type === 'form';
  /**
   *提交表单，根据两种模式不同，方法不相同
   */

  var _submit = /*#__PURE__*/function () {
    var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(firstLoad) {
      var value, finalValue;
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (isForm) {
                _context.next = 4;
                break;
              }

              value = form.getFieldsValue();
              _context.next = 12;
              break;

            case 4:
              _context.prev = 4;
              _context.next = 7;
              return form.validateFields();

            case 7:
              value = _context.sent;
              _context.next = 12;
              break;

            case 10:
              _context.prev = 10;
              _context.t0 = _context["catch"](4);

            case 12:
              if (onSubmit && valueTypeRef.current) {
                // 转化值
                // moment -> string
                // key: [value, value] -> { key:value, key: value }
                finalValue = transformKeySubmitValue(conversionSubmitValue(value, dateFormatter, valueTypeRef.current), transformKeyRef.current);
                onSubmit(finalValue, firstLoad);
              }

            case 13:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[4, 10]]);
    }));

    return function submit(_x) {
      return _ref4.apply(this, arguments);
    };
  }();

  useEffect(function () {
    if (!formRef) {
      return;
    }

    if (typeof formRef === 'function') {
      formRef(form);
    }

    if (formRef && typeof formRef !== 'function') {
      // eslint-disable-next-line no-param-reassign
      formRef.current = _objectSpread(_objectSpread({}, form), {}, {
        submit: function submit() {
          _submit(false);

          form.submit();
        }
      });
    }
  }, []);
  useDeepCompareEffect(function () {
    if (counter.proColumns.length < 1) {
      return;
    }

    var tempMap = {};
    var transformKeyMap = {};
    counter.proColumns.forEach(function (item) {
      var key = item.key,
          dataIndex = item.dataIndex,
          index = item.index,
          valueType = item.valueType,
          search = item.search,
          hideInSearch = item.hideInSearch;
      warningOnce(typeof hideInSearch !== 'boolean', "'hideInSearch' will be deprecated, please use 'search'"); // 以key为主,理论上key唯一

      var finalKey = genColumnKey(key || dataIndex, index); // 如果是() => ValueType

      var finalValueType = typeof valueType === 'function' ? valueType(item, type) : valueType;
      tempMap[finalKey] = finalValueType;
      transformKeyMap[finalKey] = typeof search === 'boolean' || !search ? undefined : function (value, fieldName, target) {
        return search === null || search === void 0 ? void 0 : search.transform(value, fieldName, target);
      };
    }); // 触发一个 submit，之所以这里触发是为了保证 value 都被 format了

    if (!valueTypeRef.current && type !== 'form') {
      // 下面才去赋值，所以用 setTimeout 延时一下，略微性能好一点
      setTimeout(function () {
        _submit(true);
      }, 0);
    }

    valueTypeRef.current = tempMap;
    transformKeyRef.current = transformKeyMap;
  }, [counter.proColumns]);

  var _useContext = useContext(_ConfigProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var columnsList = counter.proColumns.filter(function (item) {
    var valueType = item.valueType;

    if ((item.hideInSearch || item.search === false) && type !== 'form') {
      return false;
    }

    if (type === 'form' && item.hideInForm) {
      return false;
    }

    if (valueType !== 'index' && valueType !== 'indexBorder' && valueType !== 'option' && (item.key || item.dataIndex)) {
      return true;
    }

    return false;
  }).sort(function (a, b) {
    if (a && b) {
      return (b.order || 0) - (a.order || 0);
    }

    if (a && a.order) return -1;
    if (b && b.order) return 1;
    return 0;
  });
  var domList = columnsList.map(function (item, index) {
    var _item$dataIndex;

    return proFormItemRender({
      isForm: isForm,
      formInstance: formInstanceRef.current,
      item: _objectSpread({
        key: ((_item$dataIndex = item.dataIndex) === null || _item$dataIndex === void 0 ? void 0 : _item$dataIndex.toString()) || index,
        index: index
      }, item),
      type: type,
      intl: intl
    });
  }).filter(function (item) {
    return !!item;
  });
  var className = getPrefixCls('pro-table-search');
  var formClassName = getPrefixCls('pro-table-form');

  var _getFormCompetent = getFormCompetent(isForm, searchConfig),
      Competent = _getFormCompetent.Competent,
      competentName = _getFormCompetent.competentName; // 传给每个表单的配置，理论上大家都需要


  var loadingProps = {
    submitter: {
      submitButtonProps: {
        loading: submitButtonLoading
      }
    }
  };
  return /*#__PURE__*/React.createElement("div", {
    className: classNames(className, (_classNames = {}, _defineProperty(_classNames, formClassName, isForm), _defineProperty(_classNames, getPrefixCls("pro-table-search-".concat(competentName)), true), _classNames))
  }, /*#__PURE__*/React.createElement(Competent, _extends({}, loadingProps, getFromProps(isForm, searchConfig, competentName), formConfig, {
    form: form,
    onValuesChange: function onValuesChange(change, all) {
      forceUpdate();

      if (formConfig.onValuesChange) {
        formConfig.onValuesChange(change, all);
      }
    },
    onReset: function onReset() {
      if (_onReset) {
        var value = form.getFieldsValue();

        _onReset(value);
      }
    },
    onFinish: function onFinish() {
      _submit(false);
    },
    initialValues: formConfig.initialValues
  }), domList));
};

export default FormSearch;
